extensions/amp-web-push/0.1/web-push-service.js

```diff
@@ -0,0 +1,697 @@
+import {installStylesForShadowRoot} from '../../../src/shadow-embed';
+import {openWindowDialog} from '../../../src/dom';
+import {TAG, NotificationPermission} from './vars';
+import {WebPushWidgetVisibilities} from './amp-web-push-widget';
+import {parseJson} from '../../../src/json';
+
+export class WebPushService {
+
+  /*
+    In environments where pop ups aren't supported, the AMP page is redirected
+    to a lightweight "permission dialog" page on the canonical origin. After
+    permissions are granted, the page is redirected back to the AMP page.
+
+    This describes the URL query parameter to add to the redirect back to the
+    AMP page, so we know to resume the subscription process.
+
+    We use the History API after to remove this fragment from the URL without
+    affecting the page state.
+   */
+  static get PERMISSION_POPUP_URL_FRAGMENT() {
+    return 'amp-web-push-subscribing=yes';
+  };
+
+  static get AMP_VERSION_INITIAL() {
+    return 1;
+  }
+
+  /**
+   * @param {!../../../src/service/ampdoc-impl.AmpDoc} ampdoc
+   */
+  constructor(ampdoc) {
+    this.ampdoc = ampdoc;
+
+    // Enable our AMP extension for development and test environments
+    this.enableAmpExperimentForDevelopment_();
+    // On all other environments, error if experiment is not enabled
+    this.ensureAmpExperimentEnabled_();
+
+    // Install styles.
+    if (ampdoc.isSingleDoc()) {
+      const root = /** @type {!Document} */ (ampdoc.getRootNode());
+      installStyles(root, CSS, () => { }, false, TAG);
+    } else {
+      const root = /** @type {!ShadowRoot} */ (ampdoc.getRootNode());
+      installStylesForShadowRoot(root, CSS, false, TAG);
+    }
+
+    /**
+     * Resolved when the service is fully initialized.
+     * @const @private {Promise}
+     */
+    this.initializePromise_ = ampdoc.whenReady().then(() => {
+      return this.initialize_();
+    });
+
+    /** @private {Object} */
+    this.config_ = {
+      helperIframeUrl: null,
+      permissionDialogUrl: null,
+      serviceWorkerUrl: null,
+      debug: null,
+    };
+
+    /** @private {boolean} */
+    this.debug_ = false;
+
+    /** @private {./iframe.IFrame} */
+    this.iframe_ = null;
+
+    /** @private {./window-messenger.WindowMessenger} */
+    this.frameMessenger_ = null;
+
+    /** @private {./window-messenger.WindowMessenger} */
+    this.popupMessenger_ = null;
+  }
+
+  /*
+  * Occurs when the DOM is ready to be parsed.
+  */
+  initialize_() {
+    this.log_('amp-web-push extension starting up.');
+
+    // Exit early if web push isn't supported
+    if (!this.environmentSupportsWebPush()) {
+      this.log_('Web push is not supported.');
+      return;
+    }
+
+    // Read amp-web-push configuration
+    this.config_ = this.parseConfigJson(this.getConfigAsText());
+    if (!this.config_) {
+      // An error will already be thrown from the config parsing function
+      return;
+    }
+    this.debug_ = this.config_.debug;
```

Comment:
Set `debug: getMode().development` for the main `web-push-service.js` file. 

Did not change `helper-frame.js` and `permission-dialog.js` since these files are outside the AMP framework and live on the canonical origin. `window-messenger.js` inherits the debug setting passed to it.